plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java

```diff
@@ -125,12 +126,32 @@ public static Optional<String> getFilterClause(TupleDomain<ColumnHandle> tupleDo
         return Optional.empty();
     }
 
-    private static String toPredicate(PinotColumnHandle pinotColumnHandle, Domain domain)
+    private static Optional<String> toPredicate(PinotColumnHandle pinotColumnHandle, Domain domain)
     {
         String predicateArgument = pinotColumnHandle.isAggregate() ? pinotColumnHandle.getExpression() : quoteIdentifier(pinotColumnHandle.getColumnName());
+        ValueSet valueSet = domain.getValues();
+        if (valueSet.isNone()) {
+            verify(!domain.isNullAllowed(), "IS NULL is not supported due to different null handling semantics. See https://docs.pinot.apache.org/developers/advanced/null-value-support");
+            return Optional.of(format("(%s != %s)", predicateArgument, predicateArgument));
+        }
+        if (valueSet.isAll()) {
+            verify(domain.isNullAllowed(), "IS NOT NULL is not supported due to different null handling semantics. See https://docs.pinot.apache.org/developers/advanced/null-value-support");
+            // Pinot does not support "1 = 1" syntax: see https://github.com/apache/pinot/issues/10600
+            // As a workaround, skip adding always true to conjuncts
+            return Optional.empty();
+        }
+        verify(!domain.getValues().getRanges().getOrderedRanges().isEmpty() && !domain.isNullAllowed(), "IS NULL is not supported due to different null handling semantics. See https://docs.pinot.apache.org/developers/advanced/null-value-support");
         List<String> disjuncts = new ArrayList<>();
         List<Object> singleValues = new ArrayList<>();
-        for (Range range : domain.getValues().getRanges().getOrderedRanges()) {
+        boolean invertPredicate = false;
+        if (!valueSet.isDiscreteSet()) {
+            ValueSet complement = domain.getValues().complement();
+            if (complement.isDiscreteSet()) {
+                invertPredicate = complement.isDiscreteSet();
+                valueSet = complement;
+            }
+        }
```

Comment:
nit : Maybe it could be implemented a different PR. This logic allows us to check if its for `NOT IN` types of filter condition ? Can we expose them as an API so that connectors could identify them like  `DefaultQueryBuilder` and `SqlServerClient` use similar logic. 


cc: @findepi Is it possible to expose them as an API or some `Util` method ? 


cc: @findepi Can we extract them to a Util method ?

Reply:
lmk - I can extract it, or I can wait for the pr (or even help with that) and then change this to use the utility method or api.

Reply:
I added #17189 for the api method and found (and replaced) 2 use cases outside of this pr.